phi_four.py

```diff
@@ -0,0 +1,233 @@
+from tqdm import tqdm
+import sys
+import time
+
+import numpy as np
+import matplotlib.pyplot as plt
+
+import torch
+import torch.optim as optim
+
+from norm_flow_pytorch import NormalisingFlow, shifted_kl
+
+class array_mapping(object):
```

Comment:
I agree this can be slimmed down. We can discuss tomorrow but I think a single function is sufficient. I wrote a proposal one based on your code and check they do the same thing